File: verifier.py
#!/usr/bin/env python3
"""
RightsDocket Verifier — Reference Implementation (Python)

Usage:
    python verifier.py <path-to-assertion.json>

Exit codes:
    0 = VALID, 1 = INVALID, 2 = ERROR

Dependencies: pip install pynacl
Spec: https://www.rightsdocket.com/verify
"""
import base64
import json
import os
import sys
import urllib.request
from typing import Any

import nacl.signing
import nacl.exceptions

JWKS_URL = os.getenv(
    'RIGHTSDOCKET_JWKS_URL',
    'https://www.rightsdocket.com/.well-known/jwks.json'
)


def b64url_decode(s: str) -> bytes:
    s += '=' * (-len(s) % 4)
    return base64.urlsafe_b64decode(s)


def canonicalize(obj: Any) -> str:
    """Minimal RFC 8785 JCS-compatible canonicalization."""
    if obj is None or isinstance(obj, (bool, int, float, str)):
        return json.dumps(obj, separators=(',', ':'), ensure_ascii=False)
    if isinstance(obj, list):
        return '[' + ','.join(canonicalize(x) for x in obj) + ']'
    if isinstance(obj, dict):
        items = sorted(obj.items())
        return '{' + ','.join(
            json.dumps(k, separators=(',', ':')) + ':' + canonicalize(v)
            for k, v in items
        ) + '}'
    raise TypeError(f'Unsupported type: {type(obj)}')


def fetch_jwks() -> dict:
    with urllib.request.urlopen(JWKS_URL, timeout=10) as r:
        return json.loads(r.read())


def verify(assertion_path: str) -> dict:
    with open(assertion_path) as f:
        assertion = json.load(f)

    kid = assertion.get('key_id') or assertion.get('kid')
    sig_b64 = assertion.get('signature')
    payload = assertion.get('payload')
    tsr = assertion.get('timestamp') or assertion.get('tsr')

    if not (kid and sig_b64 and payload):
        return {'valid': False, 'reason': 'assertion missing key_id, signature, or payload'}

    jwks = fetch_jwks()
    jwk = next((k for k in jwks['keys'] if k.get('kid') == kid), None)
    if not jwk:
        return {'valid': False, 'reason': f'key_id {kid} not found in published JWKS'}
    if jwk.get('kty') != 'OKP' or jwk.get('crv') != 'Ed25519':
        return {'valid': False, 'reason': f'unexpected key type for {kid}'}

    pub_bytes = b64url_decode(jwk['x'])
    verify_key = nacl.signing.VerifyKey(pub_bytes)

    canonical = canonicalize(payload)
    signature = b64url_decode(sig_b64)

    try:
        verify_key.verify(canonical.encode('utf-8'), signature)
        sig_valid = True
    except nacl.exceptions.BadSignatureError:
        sig_valid = False

    has_timestamp = bool(tsr) and len(tsr) > 100

    return {
        'valid': sig_valid and has_timestamp,
        'signature_valid': sig_valid,
        'timestamp_present': has_timestamp,
        'key_id': kid,
        'fingerprint_jws': jwk.get('x5t#S256'),
        'reason': 'OK' if (sig_valid and has_timestamp) else (
            'missing or empty timestamp' if sig_valid else 'signature did not verify'
        ),
    }


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print('Usage: python verifier.py <assertion.json>', file=sys.stderr)
        sys.exit(2)

    try:
        result = verify(sys.argv[1])
        print(json.dumps(result, indent=2))
        sys.exit(0 if result['valid'] else 1)
    except Exception as e:
        print(f'ERROR: {e}', file=sys.stderr)
        sys.exit(2)

File: verifier.mjs
#!/usr/bin/env node
/**
 * RightsDocket Verifier — Reference Implementation
 *
 * Usage:
 *   node verifier.mjs <path-to-assertion.json>
 *
 * Exit codes:
 *   0 = VALID, 1 = INVALID, 2 = ERROR
 *
 * Dependencies: npm install @noble/ed25519 @noble/hashes
 * Spec: https://www.rightsdocket.com/verify
 */

import * as ed from '@noble/ed25519';
import { sha512 } from '@noble/hashes/sha512';
import { readFileSync } from 'node:fs';

ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

const JWKS_URL = process.env.RIGHTSDOCKET_JWKS_URL
  ?? 'https://www.rightsdocket.com/.well-known/jwks.json';

function b64urlToBytes(b64url) {
  const b64 = b64url.replace(/-/g, '+').replace(/_/g, '/');
  const padded = b64 + '='.repeat((4 - (b64.length % 4)) % 4);
  return new Uint8Array(Buffer.from(padded, 'base64'));
}

function canonicalize(obj) {
  if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
  if (Array.isArray(obj)) return '[' + obj.map(canonicalize).join(',') + ']';
  const keys = Object.keys(obj).sort();
  return '{' + keys.map(k => JSON.stringify(k) + ':' + canonicalize(obj[k])).join(',') + '}';
}

async function fetchJwks() {
  const res = await fetch(JWKS_URL);
  if (!res.ok) throw new Error(`JWKS fetch failed: ${res.status}`);
  return res.json();
}

async function verify(assertionPath) {
  const raw = readFileSync(assertionPath, 'utf8');
  const assertion = JSON.parse(raw);
  const kid = assertion.key_id ?? assertion.kid;
  const sigB64 = assertion.signature;
  const payload = assertion.payload;
  const tsr = assertion.timestamp ?? assertion.tsr;

  if (!kid || !sigB64 || !payload) {
    throw new Error('assertion missing key_id, signature, or payload');
  }

  const jwks = await fetchJwks();
  const jwk = jwks.keys.find(k => k.kid === kid);
  if (!jwk) {
    return { valid: false, reason: `key_id ${kid} not found in published JWKS` };
  }
  if (jwk.kty !== 'OKP' || jwk.crv !== 'Ed25519') {
    return { valid: false, reason: `unexpected key type for ${kid}` };
  }

  const pubKey = b64urlToBytes(jwk.x);
  const canonical = canonicalize(payload);
  const message = new TextEncoder().encode(canonical);
  const signature = b64urlToBytes(sigB64);

  const valid = await ed.verifyAsync(signature, message, pubKey);
  const hasTimestamp = !!tsr && tsr.length > 100;

  return {
    valid: valid && hasTimestamp,
    signature_valid: valid,
    timestamp_present: hasTimestamp,
    key_id: kid,
    fingerprint_jws: jwk['x5t#S256'],
    reason: valid && hasTimestamp ? 'OK' : (valid ? 'missing or empty timestamp' : 'signature did not verify'),
  };
}

const path = process.argv[2];
if (!path) {
  console.error('Usage: node verifier.mjs <assertion.json>');
  process.exit(2);
}

verify(path)
  .then(result => {
    console.log(JSON.stringify(result, null, 2));
    process.exit(result.valid ? 0 : 1);
  })
  .catch(err => {
    console.error('ERROR:', err.message);
    process.exit(2);
  });
